fix(computedFrom): add initialValue + throw Error in case of not sync emit

dmorosinotto opened the pull request:
Hi @eneajaho I've finally find the time to make some changes to computedFrom as prompted in #37 

1. My first commit add some more **tests** to your current implementation to prove the issue that I have with _"spurious"_ sync emit of `null` or initial input in case where observable don't have a valid sync emitted value. 

You document those cases as `tricky part` in the docs, but I believe that we can handle it differently throwing an error as the "native" Angular `toSignal` do, so the developer knows where his code will generate the problem and decide how to fix it!

2. I changed the `computedFrom` implementation and tests to handle the `initialValue: Output` passed in options object as last parameter (the options object can be used even to pass the `injector` as before) - I rewrote the test that use the new behavior.

3. In the last commit I updated the `docs` to explain these new behaviors: initialValue + throwing Error.

I hope you can review this PR and give me some feedback, because in the old Issue we didn't have the opportunity to discuss much about how to proceed, and so I decided to complete the code and tests to have a discussion base, and maybe fix or improve it even more to have a  _"better"_ `computedFrom` 🤓

dmorosinotto commented:
Hi @eneajaho sorry to directly push my code into PR without previous discussion in the issue #37 

But if you agree we can discuss/refine the idea about computedFrom here directly on the code side 😉

I added _“many comments”_ to explain my ideas and motivation, obviusly we’ll remove them befote merge, but let’s start the discussion and give me any feedback about my changes, even what you think about the breaking behaviour of throwing error 👀

dmorosinotto commented:
Hi @eneajaho , @nartc I know you are both very busy following all the latest Angular news & community, but is there a chance someone would take a look at this PR and gives me any feedback or maybe start a discussion on what you don't like, to make it progress and eventually the changes needed before merging it?

Thanks in advance and sorry to bother you, I really just want to be collaborative... not criticism!

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
```diff
@@ -71,7 +73,8 @@ let c = computedFrom(
 				of(a + b).pipe(delay(1000)) // delay the emission of the combined value by 1 second for demonstration purposes
-		)
+		),
+		{ initialValue: 42 } // change the initial value of the resulting signal
```
We don't need this here as signals have an initial value by default, so it's not the case. 
```suggestion
```

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
````diff
@@ -87,28 +90,28 @@ setTimeout(() => {
 
 ```ts
--[1, 2] - // initial value
+42 - // initial value
````
Revert this please

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
```diff
@@ -71,7 +73,8 @@ let c = computedFrom(
 				of(a + b).pipe(delay(1000)) // delay the emission of the combined value by 1 second for demonstration purposes
-		)
+		),
+		{ initialValue: 42 } // change the initial value of the resulting signal
```
I'd like to have this here as is. And add another example where we use initial value to showcase how it changes the output.

dmorosinotto replied:
If I remove the { initialValue: 42 } this sample will THROW becouse it's exactly the case where the input + delay operator generate "late async" emit , so with the new implementation, you don't get the "old spurious" initial emit of array [1,2] (that is what I really don't like because it's totally un-expected (and maybe error prone) with an inferred output as Signal<number>.

This is why I introduced { initialValue: 42 } in the sample, and adjusted the output expected in the console.

Let's me know if you agree, or if you prefer to have a different sample showing the use-case using only inputs + pipe Operator, and keeping this one to show the initialValue to prevent throwing error... 👀

eneajaho replied:
Let's have both cases. The one that throws and the new one that doesn't.

dmorosinotto replied:
Ok @eneajaho I added the first sample without params so it'll throw error, and then show the two possible solutions with initialValue or with startWith. 
Let's me know what you think?

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
```diff
@@ -87,28 +90,28 @@ setTimeout(() => {
 
-As we can see, the first value will not be affected by the rxjs operators, because they are asynchronous and the first value is emitted synchronously.
-In order to change the first value, we can use startWith operator.
+As we can see, we passed an `initialValue` as third argument, to prevent _throwing error_ in case of observable that will **not has sync value** becouse they emit later their values.
```
```suggestion
As we can see, we passed an `initialValue` as the third argument, to prevent returning `null` in case of an observable that will **not have a sync value** because they emit later their values.
```

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
```diff
@@ -151,15 +154,13 @@ console.log(combinedObject()); // { page: 1, filters: { name: 'John' } }
 
 :::note[Tricky part]
-For `Observable`s that don't emit synchronously, `computedFrom` will give us null as the initial value for the `Observable`s.
+For `Observable`s that don't emit synchronously `computedFrom` will **throw and error** forcing you to fix this situation either passing an `initialValue` in third argument, or using `startWith` operetor to force observable to have a sync starting value.
```
```suggestion
For `Observable`s that don't emit synchronously `computedFrom` will **throw an error** forcing you to fix this situation either by passing an `initialValue` in the third argument, or using `startWith` operator to force observable to have a sync starting value.
```

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
```diff
@@ -151,15 +154,13 @@ console.log(combinedObject()); // { page: 1, filters: { name: 'John' } }
-const combined = computedFrom([page$, filters$]);
-
-console.log(combined()); // [null, { name: 'John' }]
+const combined = computedFrom([page$, filters$]); // 👈 WILL THROW AN ERROR!!!
```
```suggestion
const combined = computedFrom([page$, filters$]); // 👈 will throw an error!! 💥
```

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
````diff
@@ -173,9 +174,20 @@ const combined = computedFrom([
 ```
 
-### 4. Use it outside of an injection context
+### 4. Use of options initialValue
````
```suggestion
### 4. Using initialValue param
```

eneajaho commented on libs/ngxtension/computed-from/src/computed-from.spec.ts:
```diff
@@ -25,6 +25,7 @@ import {
 	switchMap,
 } from 'rxjs';
 import { computedFrom } from './computed-from';
+import exp = require('constants');
```
Is this needed?

dmorosinotto replied:
No it was an auto-import mistake will remove it

eneajaho commented on libs/ngxtension/computed-from/src/computed-from.spec.ts:
```diff
@@ -58,6 +59,30 @@ describe(computedFrom.name, () => {
+				const late = of(1).pipe(delay(1000)); // late emit after 1s
+				expect(() => {
+					const s = computedFrom([late]);
+				}).toThrowError(/requireSync/i); //THROW ERROR NG0601 DUE TO toSignal + requireSync: true
```
```suggestion
				}).toThrowError(/requireSync/i); // Throw error NG0601 due to `toSignal` + `requireSync: true`
```

eneajaho commented on libs/ngxtension/computed-from/src/computed-from.spec.ts:
```diff
@@ -58,6 +59,30 @@ describe(computedFrom.name, () => {
+				expect(() => {
+					const s = computedFrom([late]);
+				}).toThrowError(/requireSync/i); //THROW ERROR NG0601 DUE TO toSignal + requireSync: true
+				//THIS WILL PREVENT OLD "SPURIOUS SYNC EMIT" OF null OR Input ([], {}) THAT CAN CAUSE TS RUNTIME ERRORS
```
Can you use normal lowercase sentences, please? 

Uppercase is not easy to read.

eneajaho commented on libs/ngxtension/computed-from/src/computed-from.spec.ts:
```diff
@@ -58,6 +59,30 @@ describe(computedFrom.name, () => {
+				// expect(s()[0].toFixed(2)).toEqual('1.00'); //HERE WE CAN CALL s()[0].toFixed(2) <-- THIS WILL WORK
+			});
+		});
+		it(`MD FIX for Observables that don't emit synchronously, you can pass options.initialValue TO PREVENT ERROR`, fakeAsync(() => {
```
What is MD ?

dmorosinotto replied:
Will remove all of it

eneajaho commented on libs/ngxtension/computed-from/src/computed-from.ts:
```diff
@@ -59,53 +86,82 @@ export function computedFrom<Input extends object, Output = Input>(
+	let injector = options?.injector;
 	injector = assertInjector(computedFrom, injector);
+	/* try { //CUSTOM ERROR HANDLING FOR computedFrom */
+	//IF YOU PASS options.initialValue RETURN Signal<Output> WITHOUT ANY PROBLEM EVEN IF sources Observable ARE ASYNC (LATE EMIT) -> OUTPUT SIGNAL START EMITING SYNC initialValue!
```
Can you convert the text to lowercase, please?

dmorosinotto replied:
Sorry forgot of this uppercase in the code - now I fixed it.
 @nartc & @eneajaho please let's me know if what you think of the new behavior I introduced?

eneajaho commented on libs/ngxtension/computed-from/src/computed-from.ts:
```diff
@@ -59,53 +86,82 @@ export function computedFrom<Input extends object, Output = Input>(
+			e.message.includes('NG601') ||
+			e.code == 601
+		)
+			console.warn(
```
@nartc What do you think about this one? 

I think we will need sth like this.

dmorosinotto replied:
@eneajaho when you ask me to open a new issue to discuss try catch handle error are you referring to those commented lines where I suggest a way to handle the error due to `requireSync` to customize the message for `computedForm`?

eneajaho commented on libs/ngxtension/computed-from/src/computed-from.ts:
```diff
@@ -59,53 +86,82 @@ export function computedFrom<Input extends object, Output = Input>(
+} {
+	if (!args || !args.length || typeof args[0] !== 'object')
+		//VALID EVEN FOR ARRAY
+		throw new TypeError('computedFrom need sources');
```
```suggestion
		throw new TypeError('computedFrom needs sources');
```

eneajaho commented on libs/ngxtension/computed-from/src/computed-from.ts:
```diff
@@ -59,53 +86,82 @@ export function computedFrom<Input extends object, Output = Input>(
+		throw new TypeError('computedFrom need sources');
+	const hasOperator = typeof args[1] === 'function';
+	if (args.length == 3 && !hasOperator)
+		throw new TypeError('computedFrom need pipebale operator as second arg');
```
```suggestion
		throw new TypeError('computedFrom needs pipeable operator as a second argument');
```

eneajaho commented:
Hi @dmorosinotto 

It looks good to me. It doesn't look breaking to me. There are some comments I'd like you to review. 

And we need to discuss the error-handling part with @nartc . 

Thanks for your patience.

dmorosinotto commented:
Hi @eneajaho thanks for your feedback reviewing my code, today I'm quite busy at work, but I'll take all your suggestions to update my code this evening (after 21:00 CEST) 
Just to give you some quick response (than I'll detail it in context of your comments)
- I'll remove/lower all UPPERCASE comments
- about the reverted initialValue sample where in your original implementation we get the [1,2] emission, it's not possible becouse in case of "async later emit" the new code will `throwError` becouse I added `pipe(delay(1000))` 
- about the `MD & MD FIX` in the tests title it was just my signature to easly find/quick identify where I changed the behavior / breacking change respect to your older behavior, I'll *absolutely remove* `MD` from the final code!
-  I don't know where I added the `exp = required("...")` probably some wrong auto-import, obviously I'll remove it

dmorosinotto commented:
OK @eneajaho I try to fix the typos and uppercase tests comments, hope this will made it more readable and conform to the repo standards.

Waiting for your feedback and @nartc opinion about the new `throwing error` behavior 👀

dmorosinotto commented:
@eneajaho , @nartc I fixed the CI errors - if you want to review the code and give me some feedback about what you think of the current "throwing error" behavior?  

BTW: I had to do some "try&error" to make a commit passing CI, but running local test always pass even with previous failing CI commits, I suspect that the CI is using some stricter linting rules! Can we align those so anyone can verify all code & test works as expected from CI before commit? Thanks

PS: hope to meet you IRL at NGPoland next week, to discuss this or other Angular stuff 😉

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
````diff
@@ -84,31 +86,46 @@ setTimeout(() => {
 ```
 
-The console log will be:
+This will _throw an error_ because the operation pipeline used will produce an observable that will **not have a sync value** because they emit later their values, so the resultig `c` signal doesn't have an initial value, and this caused the error.
````
```suggestion
This will _throw an error_ because the operation pipeline will produce an observable that will **not have a sync value** because they emit their values later on, so the resulting `c` signal doesn't have an initial value, and this causes the error.
```

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
````diff
@@ -84,31 +86,46 @@ setTimeout(() => {
+);
+```
+
+This will works, and you can copy the above example inside an Angular constructor and see the result in the console:
````
```suggestion
This works, and you can copy the above example inside an Angular constructor and see the result in the console:
```

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
```diff
@@ -84,31 +86,46 @@ setTimeout(() => {
 
-As we can see, the first value will not be affected by the rxjs operators, because they are asynchronous and the first value is emitted synchronously.
-In order to change the first value, we can use startWith operator.
+Another way to solve the same problem, is using the `startWith` operator in the pipe to force the observable to have a starting value like below.
```
```suggestion
Another way to solve this problem is using the `startWith` rxjs operator in the pipe to force the observable to have a starting value like below.
```

eneajaho commented on docs/src/content/docs/utilities/Signals/computed-from.md:
```diff
@@ -84,31 +86,46 @@ setTimeout(() => {
 	pipe(
 		switchMap(([a, b]) => of(a + b).pipe(delay(1000))),
-		startWith(0) // change the first value
+		startWith(0) // 👈 change the starting value (sync emit)
```
```suggestion
		startWith(0) // 👈 change the starting value (emits synchronously)
```

eneajaho commented:
Hi @dmorosinotto 
Thanks, it looks good now. 

Regarding try catch error handling, would you mind opening an issue, so we can tackle it later and get this PR merged? 

Also, please can you fix those other change requests? And after that it should be good to be merged.

dmorosinotto commented:
Hi @eneajaho I've added a commit to fix remaining typo in docs.
 
And just for clarification about the issue that you want me to open - are you referring to commented lines in `computedFrom.ts` L113-128 where I suggest a way to handle&customize error message due to `requiredSync`?

nartc commented:
I hate you guys for having fun at ngpoland conf without me so I rage merged this PR

eneajaho commented:
Oh, I'm so sorry for you @nartc  🫶
